matrix-sdk/src/main/java/org/matrix/androidsdk/rest/client/MediaScanRestClient.java

```
@@ -15,47 +15,152 @@
+                    if (e.mStatus == 404) {
+                        // On 404 consider the public key is not available, so do not encrypt body
+                        mMxStore.setAntivirusServerPublicKey("");
+
+                        callback.onSuccess("");
+                    } else {
+                        super.onMatrixError(e);
+                    }
+                }
+            }));
+        }
+    }
+
+    /**
+     * Reset Antivirus server public key on cache
+     */
+    public void resetServerPublicKey() {
+        if (mMxStore != null) {
+            mMxStore.setAntivirusServerPublicKey(null);
+        }
+    }
+
     /**
      * Scan an unencrypted file.
      *
      * @param domain   the server name extracted from the matrix content uri
-     * @param mediaId   the media id extracted from the matrix content uri
+     * @param mediaId  the media id extracted from the matrix content uri
      * @param callback on success callback containing a MediaScanResult object
      */
     public void scanUnencryptedFile(final String domain, final String mediaId, final ApiCallback<MediaScanResult> callback) {
-
         mApi.scanUnencrypted(domain, mediaId).enqueue(new DefaultRetrofit2CallbackWrapper<>(callback));
     }
 
     /**
      * Scan an encrypted file.
      *
      * @param encryptedMediaScanBody the encryption information required to decrypt the content before scanning it.
-     * @param callback on success callback containing a MediaScanResult object
+     * @param callback               on success callback containing a MediaScanResult object
      */
     public void scanEncryptedFile(final EncryptedMediaScanBody encryptedMediaScanBody, final ApiCallback<MediaScanResult> callback) {
+        // Encrypt encryptedMediaScanBody if the server support it
+        getServerPublicKey(new SimpleApiCallback<String>(callback) {
+            @Override
+            public void onSuccess(String info) {
```

I would rename the param "info" with "serverPublicKey" (or "curve25519PublicKey")

---

The case where this key is null is not handled correctly.
We should return an error in this case without triggering a scan request